File: data_structures/week6_binary_search_trees/2_is_bst/is_bst.py
import threading
from typing import List, Union


def is_bst(  # noqa: CFQ004, C901
    tree: List[List[int]],
    *,
    index: int = 0,
    min_value: Union[int, float] = float("-inf"),
    max_value: Union[int, float] = float("inf"),
) -> bool:
    if not tree:
        return True

    if index == -1:
        return True

    if tree[index][0] < min_value or tree[index][0] >= max_value:
        return False

    left_is_bst = is_bst(
        tree,
        index=tree[index][1],
        min_value=min_value,
        max_value=tree[index][0],
    )

    right_is_bst = is_bst(
        tree,
        index=tree[index][2],
        min_value=tree[index][0],
        max_value=max_value,
    )

    return left_is_bst and right_is_bst


def main() -> None:
    nodes = int(input().strip())
    tree = [list(map(int, input().strip().split())) for _ in range(nodes)]
    print("CORRECT" if is_bst(tree) else "INCORRECT")


if __name__ == "__main__":
    threading.Thread(target=main).start()

File: data_structures/week6_binary_search_trees/2_is_bst/is_bst.cpp
#include <algorithm>
#include <iostream>
#include <vector>

namespace is_bst {
struct Node {
  int key;
  int left;
  int right;

  Node() : key(0), left(-1), right(-1) {}
  Node(int key_, int left_, int right_) : key(key_), left(left_), right(right_) {}
};

bool is_bst(const std::vector<Node>& tree, int idx = 0, int min = -2147483648, int max = 2147483647) {
  if (tree.size() == 0) return true;
  if (idx == -1) return true;
  if (tree[idx].key < min || tree[idx].key >= max) return false;

  bool left_is_bst = is_bst(tree, tree[idx].left, min, tree[idx].key);
  bool right_is_bst = is_bst(tree, tree[idx].right, tree[idx].key, max);

  return left_is_bst && right_is_bst;
}
}  // namespace is_bst

int main() {
  int nodes;
  std::cin >> nodes;

  std::vector<is_bst::Node> tree;
  for (int i = 0; i < nodes; ++i) {
    int key, left, right;
    std::cin >> key >> left >> right;
    tree.push_back(is_bst::Node(key, left, right));
  }
  std::cout << (is_bst::is_bst(tree) ? "CORRECT" : "INCORRECT") << std::endl;
  return 0;
}
